reclaimspace/README.md

```diff
@@ -0,0 +1,159 @@
+
+* **Controller Service**: The Controller plugin MAY implement this RPC.
+* **Node Service**: The Node plugin MAY implement this RPC.
+
+```protobuf
+syntax = "proto3";
+package reclaimspace;
+
+import "google/protobuf/descriptor.proto";
+
+option go_package = ".;reclaimspace";
+
+extend google.protobuf.FieldOptions {
+  // Indicates that a field MAY contain information that is sensitive
+  // and MUST be treated as such (e.g. not logged).
+  bool reclaimspace_secret = 1099;
+}
+
+// Controller holds the RPC method for running discard operations on the
+// inactive (not required to be staged/published) volume.
+service Controller {
+  // ControllerReclaimSpace is a procedure that gets called on the CSI
+  // Controller.
+  rpc ControllerReclaimSpace (ControllerReclaimSpaceRequest)
+  returns (ControllerReclaimSpaceResponse) {}
+}
+
+// Node holds the RPC method for running discard operations on the active
+// (staged/published) volume.
+service Node {
+  // NodeReclaimSpace is a procedure that gets called on the CSI NodePlugin.
+  rpc NodeReclaimSpace (NodeReclaimSpaceRequest)
+  returns (NodeReclaimSpaceResponse) {}
+}
+```
+
+### ControllerReclaimSpace
+
+```protobuf
+// ControllerReclaimSpaceRequest holds the required information to enable
+// replication on a volume.
+message ControllerReclaimSpaceRequest {
+  // The ID of the volume. This field is REQUIRED.
+  string volume_id = 1;
+
+  // Plugin specific parameters passed in as opaque key-value pairs.
+  map<string, string> parameters = 2;
+  // Secrets required by the plugin to complete the request.
+  map<string, string> secrets = 3 [(reclaimspace_secret) = true];
+}
+
+// ControllerReclaimSpaceResponse holds the information about the result of the
+// ControllerReclaimSpaceRequest call.
+message ControllerReclaimSpaceResponse {
+  // This field is OPTIONAL. This allows the SP to inform the CO about the
+  // storage consumption before the ReclaimSpace operation was executed.
+  StorageConsumption pre_usage = 1;
+
+  // This field is OPTIONAL. This allows the SP to inform the CO about the
+  // storage consumption after the ReclaimSpace operation was executed.
+  StorageConsumption post_usage = 2;
+}
+
+// StorageConsumption contains the usage in bytes.
+message StorageConsumption {
+  // This field is REQUIRED. usage_bytes contains the consumed storage in
+  // bytes.
+  int64 usage_bytes = 1;
+}
+```
+
+### NodeReclaimSpace
+
+```protobuf
+// NodeReclaimSpaceRequest holds the required information to enable
+// replication on a volume.
+message NodeReclaimSpaceRequest {
+  // The ID of the volume. This field is REQUIRED.
+  string volume_id = 1;
+
+  // The path on which volume is available. This field is REQUIRED.
+  // This field overrides the general CSI size limit.
+  // SP SHOULD support the maximum path length allowed by the operating
+  // system/filesystem, but, at a minimum, SP MUST accept a max path
+  // length of at least 128 bytes.
+  string volume_path = 2;
```

Comment:
is this a target path?

Reply:
yes, probably. I'll check the CSI spec again for the name of the variable.

Reply:
https://github.com/container-storage-interface/spec/blob/master/spec.md#nodeexpandvolume uses `volume_path` too, I'll keep it here too.

Reply:
>The staging_target_path field is not required, for backwards compatibility, but the CO SHOULD supply it. Plugins can use this field to determine if volume_path is where the volume is published or staged, and setting this field to non-empty allows plugins to function with less stored state on the node.

it says stagingPath is not required. do we need both or have only volume_path here?

Reply:
`volume_path` is most likely sufficient, but maybe some storage providers require the staging path. It is the same for NodeExpandVolume.